src/Features/Core/Portable/SplitOrMergeIfStatements/AbstractMergeIfStatementsCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,86 @@
+
+        protected abstract Task<bool> CanBeMergedAsync(
+            Document document, SyntaxNode ifLikeStatement, CancellationToken cancellationToken);
+
+        protected abstract SyntaxNode GetChangedRoot(Document document, SyntaxNode root, SyntaxNode ifLikeStatement);
+
+        public sealed override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var root = await context.Document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
+            var node = root.FindNode(context.Span, getInnermostNodeForTie: true);
+
+            if (IsApplicableSpan(node, context.Span, out var ifLikeStatement))
+            {
+                var syntaxFacts = context.Document.GetLanguageService<ISyntaxFactsService>();
+                var syntaxKinds = context.Document.GetLanguageService<ISyntaxKindsService>();
+
+                if (await CanBeMergedAsync(context.Document, ifLikeStatement, context.CancellationToken).ConfigureAwait(false))
+                {
+                    context.RegisterRefactoring(
+                        CreateCodeAction(
+                            c => RefactorAsync(context.Document, context.Span, c),
+                            syntaxFacts.GetText(syntaxKinds.IfKeyword)));
+                }
+            }
+        }
+
+        private async Task<Document> RefactorAsync(Document document, TextSpan span, CancellationToken cancellationToken)
+        {
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var node = root.FindNode(span, getInnermostNodeForTie: true);
+
+            Contract.ThrowIfFalse(IsApplicableSpan(node, span, out var ifLikeStatement));
+
+            var newRoot = GetChangedRoot(document, root, ifLikeStatement);
+            return document.WithSyntaxRoot(newRoot);
+        }
+
+        protected static IReadOnlyList<SyntaxNode> WalkDownBlocks(ISyntaxFactsService syntaxFacts, IReadOnlyList<SyntaxNode> statements)
+        {
+            // If our statements only contain a single block, walk down the block and any subsequent nested blocks
+            // to get the real statements inside.
+
+            while (statements.Count == 1 && syntaxFacts.IsPureBlock(statements[0]))
+            {
+                statements = syntaxFacts.GetExecutableBlockStatements(statements[0]);
+            }
+
+            return statements;
+        }
+
+        protected static IReadOnlyList<SyntaxNode> WalkUpBlocks(ISyntaxFactsService syntaxFacts, IReadOnlyList<SyntaxNode> statements)
+        {
+            // If our statements are inside a block, walk up the block and any subsequent nested blocks that contain
+            // no other statements to get the topmost block.
+
+            while (statements.Count > 0 && syntaxFacts.IsPureBlock(statements[0].Parent) &&
+                   syntaxFacts.GetExecutableBlockStatements(statements[0].Parent).Count == statements.Count)
```

Comment:
nit, given how many types you state it, it might actually be cleaner to have:  `statements[0].Parent is var parent &&`

Reply:
also, you response to my comment is helpful enough to keep in the code.  atlernatively, rephrase or flesh out the existing comment.  It still isn't clear to me from reading the comment alone what this is trying to convey.

Reply:
➡️ Fixed in c9852b895b38ba50b4f51035566b7c13b21c62a7